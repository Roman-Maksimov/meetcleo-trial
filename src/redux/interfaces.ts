import { Action } from "redux";
import { sagas } from "../services/bill-service";
import Flow, { FlowList } from "./flow";
import { Saga } from "redux-saga";

export interface IFlowType {
  request: string;
  success: string;
  fail: string;
  setData: string;
  setProps: string;
  reset: string;
}

export type IState = { [key: string]: unknown };

export interface IAction extends Action<string> {
  payload?: unknown;
}

export type IError = {
  message?: string;
  response?: {
    data?: unknown;
  };
};

export interface IErrorAction extends IAction {
  error: string | IError;
  formattedError: string;
}

export interface ISetPropsAction extends IAction {
  payload?: { [key: string]: unknown };
}

export type IActionFunc<
  S extends IState = IState,
  A extends IAction = IAction
> = (state: S, action: A) => S;

export type IOnFetch<A extends IAction = IAction> = (
  action: A,
  flow: Flow
) => void;

export type IOnFail<A extends IAction = IAction> = (
  action: A,
  error: IError,
  formatterError: string
) => void;

export type IOnDone<A extends IAction = IAction> = (action: A) => void;

export type IFormatError = (error: IError, formatterError: string) => void;

export interface ISetActionsParams {
  onFetch?: IOnFetch;
  onFail?: IOnFail;
  onDone?: IOnDone;
  formatError?: IFormatError;
}

export type IFlowListSaga = { [name: string]: (action: Action) => unknown } & {
  watch?: () => void;
};

export interface IFlowState<D = {}> {
  data: D;
  fetching: boolean;
  error: string | IError | null;
  formattedError: string | null;
}

export interface IService {
  flows: FlowList;
  reducer: (state: {} | undefined, action: IAction) => unknown;
  sagas?: Saga;
}

export interface IStoreFactoryParams {
  services?: { [name: string]: IService };
  reducers?: {
    [name: string]: (state: {} | undefined, action: IAction) => unknown;
  };
  sagas?: Saga[];
  state?: {};
}
