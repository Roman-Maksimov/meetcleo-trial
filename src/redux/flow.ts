import { ActionPattern } from "@redux-saga/types";
import { v4 } from "uuid";
import { takeEvery, all, put } from "redux-saga/effects";
import {
  IAction,
  IErrorAction,
  IFlowListSaga,
  IFlowType,
  IFormatError,
  IOnDone,
  IOnFail,
  IOnFetch,
  ISetActionsParams,
  ISetPropsAction,
  IState,
} from "./interfaces";

export const createFlowType = (type: string): IFlowType => ({
  request: `${type}/REQUEST`,
  success: `${type}/SUCCESS`,
  fail: `${type}/FAIL`,
  setData: `${type}/SET_DATA`,
  setProps: `${type}/SET_PROPS`,
  reset: `${type}/RESET`,
});

export class Flow {
  private name: string;
  private type: IFlowType;
  private initialData: unknown;

  private onFetch?: IOnFetch;
  private onFail?: IOnFail;
  private onDone?: IOnDone;
  private formatError?: IFormatError;

  constructor(name: string, initialData = {}) {
    this.name = name;
    this.type = createFlowType(`${v4()}/${name}`);
    this.initialData = initialData;
  }

  getName = () => this.name;

  request = (props = {}) => ({
    ...props,
    type: this.type.request,
  });

  success = (payload: unknown) => ({
    type: this.type.success,
    payload,
  });

  fail = (error: Error, formattedError: string) => ({
    type: this.type.fail,
    error,
    formattedError,
  });

  setData = (payload: unknown) => ({
    type: this.type.setData,
    payload,
  });

  setProps = (payload: unknown) => ({
    type: this.type.setProps,
    payload,
  });

  reset = () => ({
    type: this.type.reset,
  });

  reducerRequest = (state: { [name: string]: IState }) => ({
    ...state,
    [this.name]: {
      ...state[this.name],
      data: this.initialData,
      fetching: true,
      error: null,
      formattedError: null,
    },
  });

  reducerSuccess = (state: { [name: string]: IState }, action: IAction) => ({
    ...state,
    [this.name]: {
      ...state[this.name],
      data: action.payload,
      fetching: false,
      error: null,
      formattedError: null,
    },
  });

  reducerFail = (state: { [name: string]: IState }, action: IErrorAction) => ({
    ...state,
    [this.name]: {
      ...state[this.name],
      data: this.initialData,
      fetching: false,
      error: action.error,
      formattedError: action.formattedError,
    },
  });

  reducerSetData = (state: { [name: string]: IState }, action: IAction) => ({
    ...state,
    [this.name]: {
      ...state[this.name],
      data: action.payload,
    },
  });

  reducerSetProps = (
    state: { [name: string]: IState },
    action: ISetPropsAction
  ) => ({
    ...state,
    [this.name]: {
      ...state[this.name],
      ...action.payload,
    },
  });

  reducerReset = (state: IState) => ({
    ...state,
    ...this.getInitialState(),
  });

  getInitialState = () => ({
    [this.name]: {
      data: this.initialData,
      fetching: false,
      error: null,
    },
  });

  getReducers = () => ({
    [this.type.request]: this.reducerRequest,
    [this.type.success]: this.reducerSuccess,
    [this.type.fail]: this.reducerFail,
    [this.type.setData]: this.reducerSetData,
    [this.type.setProps]: this.reducerSetProps,
    [this.type.reset]: this.reducerReset,
  });

  getSaga = () => {
    const self = this;

    function* fetch(action: IAction) {
      try {
        const payload =
          self.onFetch && typeof self.onFetch === "function"
            ? yield self.onFetch(action, self)
            : undefined;

        yield put(self.success(payload || self.initialData));
      } catch (error) {
        const payload = [];

        if (typeof error === "string") {
          payload.push(error);
        }

        if (error.message) {
          payload.push(error.message);
        }

        if (error.response) {
          payload.push(error.response.data);
        }

        let formattedError = payload.join(" \n ");

        if (self.formatError) {
          formattedError = yield self.formatError(error, formattedError);
        }

        if (self.onFail && typeof self.onFail === "function") {
          const fError = yield self.onFail(action, error, formattedError);

          if (fError !== undefined) {
            formattedError = fError;
          }
        }

        // tslint:disable-next-line
        console.error(error);

        // ToDo add auditLog here

        yield put(self.fail(error, formattedError));
      } finally {
        if (self.onDone && typeof self.onDone === "function") {
          yield self.onDone(action);
        }
      }
    }

    function* watch() {
      yield takeEvery(self.type.request, fetch);
    }

    return watch();
  };

  setActions = ({
    onFetch,
    onFail,
    onDone,
    formatError,
  }: ISetActionsParams) => {
    this.onFetch = onFetch;
    this.onFail = onFail;
    this.onDone = onDone;
    this.formatError = formatError;
  };
}

// tslint:disable-next-line
export class FlowList {
  private flows: { [name: string]: Flow } = {};
  private sagas: IFlowListSaga[] = [];
  private initialState = {};

  constructor(...flows: Flow[]) {
    this.flows = flows.reduce(
      (prev, flow) => ({
        ...prev,
        [flow.getName()]: flow,
      }),
      {}
    );
  }

  getFlow = (name: string) => this.flows[name];

  getFlows = () => this.flows;

  getInitialState = () =>
    Object.keys(this.flows).reduce(
      (prev, key) => ({
        ...prev,
        ...this.flows[key].getInitialState(),
      }),
      this.initialState
    );

  mergeInitialState = (initialState = {}) => {
    this.initialState = initialState;
  };

  getReducers = () =>
    Object.keys(this.flows).reduce(
      (prev, key) => ({
        ...prev,
        ...this.flows[key].getReducers(),
      }),
      {}
    );

  watch = (
    type: ActionPattern<IAction>,
    saga: (action: IAction) => unknown
  ) => {
    this.sagas.push({
      watch: () => {
        function* gen() {
          return yield takeEvery(type, saga);
        }

        return gen();
      },
    });
  };

  getSagas = () => {
    const self = this;

    return function* root() {
      yield all([
        ...Object.keys(self.flows).map((key) => self.flows[key].getSaga()),
        ...self.sagas.map((item) => item.watch && item.watch()),
      ]);
    };
  };
}

export default Flow;
