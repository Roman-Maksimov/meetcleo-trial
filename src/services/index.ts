import BillService, { IBillServiceState } from "./bill-service";
import CategoriesService, { ICategoriesService } from "./categories-service";

export interface IRootState {
  BillService: IBillServiceState;
  CategoriesService: ICategoriesService;
}

export default {
  BillService,
  CategoriesService,
};
