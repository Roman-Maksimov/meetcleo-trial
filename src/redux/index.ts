import * as flow from "./flow";
import _reducer from "./reducer";
import _store from "./store";

export const { Flow, FlowList } = flow;
export const reducer = _reducer;
export const store = _store;

export default {
  Flow,
  FlowList,
  reducer,
  store,
};
