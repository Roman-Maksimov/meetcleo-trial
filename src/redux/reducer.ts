import { Action } from "redux";
import { FlowList } from "./flow";
import { IAction, IState } from "./interfaces";

export default ({
  initialState = {},
  handlers = {},
  flows = new FlowList(),
}) => {
  flows.mergeInitialState(initialState);

  return (state = flows.getInitialState(), action: Action<string>) => {
    const composed: {
      [name: string]: (state: IState, action: IAction) => void;
    } = {
      ...flows.getReducers(),
      ...handlers,
    };

    return composed[action.type] ? composed[action.type](state, action) : state;
  };
};
