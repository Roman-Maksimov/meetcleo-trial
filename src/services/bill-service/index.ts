import redux from "../../redux";
import { IFlowState, IService } from "../../redux/interfaces";
import reducerFactory, { IBillServiceReducerState } from "./reducer";
import sagasFactory from "./sagas";

export const flows = new redux.FlowList(
  new redux.Flow("getAllBills", []),
  new redux.Flow("getBills", []),
  new redux.Flow("getPBills", []),
  new redux.Flow("updateBill")
);
export const reducer = reducerFactory(flows);
export const sagas = sagasFactory(flows);

const service: IService = {
  flows,
  reducer,
  sagas,
};

export interface ITransaction {
  amount: number;
  date: string;
  id: number;
}

export interface IBill {
  categoryId: number;
  iconUrl?: string;
  id: string;
  isBill: boolean;
  name: string;
  transactions: ITransaction[];
}

export interface IBillServiceState extends IBillServiceReducerState {
  getAllBills: IFlowState<IBill[]>;
  getBills: IFlowState<IBill[]>;
  getPBills: IFlowState<IBill[]>;
  updateBill: IFlowState<{ [id: string]: boolean }>;
}

export default service;
