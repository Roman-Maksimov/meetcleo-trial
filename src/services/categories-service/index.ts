import redux from "../../redux";
import { IFlowState, IService } from "../../redux/interfaces";
import reducerFactory from "./reducer";
import sagasFactory from "./sagas";

export const flows = new redux.FlowList(new redux.Flow("get", []));
export const reducer = reducerFactory(flows);
export const sagas = sagasFactory(flows);

const service: IService = {
  flows,
  reducer,
  sagas,
};

export interface ICategory {
  iconUrl?: string;
  id: number;
  name: string;
}

export interface ICategoriesService {
  get: IFlowState<{ [id: number]: ICategory }>;
}

export default service;
