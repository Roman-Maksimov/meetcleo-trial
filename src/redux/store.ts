import "regenerator-runtime/runtime";
import { createStore, applyMiddleware, compose, combineReducers } from "redux";
import createSagaMiddleware from "redux-saga";
import { default as ImplementedServices } from "../services";
import { IStoreFactoryParams } from "./interfaces";

interface IWindow {
  __REDUX_DEVTOOLS_EXTENSION__?: () => any;
}

export const storeFactory = ({
  services = {},
  reducers = {},
  sagas = [],
  state = {},
}: IStoreFactoryParams) => {
  const reducer = combineReducers(
    Object.keys(services).reduce((prev, key) => {
      const serviceReducer = services[key].reducer;
      return serviceReducer
        ? {
            ...prev,
            [key]: serviceReducer,
          }
        : prev;
    }, reducers)
  );

  const preloadedState = Object.keys(services).reduce((prev, key) => {
    const serviceFlows = services[key].flows;
    return serviceFlows
      ? {
          ...prev,
          [key]: serviceFlows.getInitialState(),
        }
      : prev;
  }, state);

  const sagaMiddlewares = createSagaMiddleware();
  const middlewares = applyMiddleware(sagaMiddlewares);

  const _store = createStore(
    reducer,
    preloadedState,
    (window as IWindow).__REDUX_DEVTOOLS_EXTENSION__
      ? compose(
          middlewares,
          (window as IWindow).__REDUX_DEVTOOLS_EXTENSION__!()
        )
      : middlewares
  );

  Object.keys(services).forEach((key) => {
    const _sagas = services[key].sagas;
    if (_sagas) {
      sagaMiddlewares.run(_sagas);
    }
  });

  sagas.forEach((saga) => {
    sagaMiddlewares.run(saga);
  });

  return _store;
};

const store = storeFactory({ services: ImplementedServices });

export default store;
