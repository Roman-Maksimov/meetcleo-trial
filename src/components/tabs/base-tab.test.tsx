import { configure, render } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import React from "react";
import { Provider } from "react-redux";
import { fakeState } from "../../../testing-data/fake-data";
import { storeFactory } from "../../redux/store";
import services from "../../services";
import BaseTabConnected from "./base-tab";

configure({ adapter: new Adapter() });

describe("BaseTab", () => {
  let store = storeFactory({ services, state: fakeState });

  beforeEach(() => {
    store = storeFactory({ services, state: fakeState });
  });

  it("should render active tab component", () => {
    const component = render(
      <Provider store={store}>
        <BaseTabConnected value={0} index={0} />
      </Provider>
    );

    expect(component).toMatchSnapshot();
  });

  it("should render inactive tab component", () => {
    const component = render(
      <Provider store={store}>
        <BaseTabConnected value={0} index={1} />
      </Provider>
    );

    expect(component).toMatchSnapshot();
  });
});
