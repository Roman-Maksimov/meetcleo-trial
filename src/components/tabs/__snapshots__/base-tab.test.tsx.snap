// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseTab should render active tab component 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "aria-labelledby": "tab-0",
      "id": "tabpanel-0",
      "role": "tabpanel",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "MuiBox-root MuiBox-root-2",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "aria-labelledby": undefined,
      "id": undefined,
      "role": undefined,
    },
    "x-attribsPrefix": Object {
      "aria-labelledby": undefined,
      "id": undefined,
      "role": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`BaseTab should render inactive tab component 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "aria-labelledby": "tab-1",
      "hidden": "",
      "id": "tabpanel-1",
      "role": "tabpanel",
    },
    "children": Array [],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "aria-labelledby": undefined,
      "hidden": undefined,
      "id": undefined,
      "role": undefined,
    },
    "x-attribsPrefix": Object {
      "aria-labelledby": undefined,
      "hidden": undefined,
      "id": undefined,
      "role": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
